loadgen/loadgen.cc

```diff
@@ -1196,6 +1072,276 @@ struct LogOutputs {
+  std::vector<loadgen::LoadableSampleSet> loadable_sets(
+      loadgen::GenerateLoadableSets(qsl, settings));
+  const LoadableSampleSet& performance_set = loadable_sets.front();
+  LoadSamplesToRam(qsl, performance_set.set);
+
+  PerformanceResult pr(IssueQueries<scenario, TestMode::PerformanceOnly>(
+      sut, settings, performance_set, sequence_gen));
+
+  sut->ReportLatencyResults(pr.latencies);
+
+  LogSummary(
+      [perf_summary = PerformanceSummary{sut->Name(), settings, std::move(pr)}](
+          AsyncSummary& summary) mutable { perf_summary.Log(summary); });
+
+  qsl->UnloadSamplesFromRam(performance_set.set);
+}
+
+/// \brief Runs the binary search mode, templated by scenario.
+template <TestScenario scenario>
+void FindPeakPerformanceMode(SystemUnderTest* sut, QuerySampleLibrary* qsl,
+                             const TestSettingsInternal& base_settings,
+                             SequenceGen* sequence_gen) {
+  LogDetail([](AsyncDetail& detail) {
+    detail("Starting FindPeakPerformance mode:");
+  });
+
+  if (scenario != TestScenario::MultiStream &&
+      scenario != TestScenario::MultiStreamFree &&
+      scenario != TestScenario::Server) {
+    LogDetail([unsupported_scenario = ToString(scenario)](AsyncDetail& detail) {
+      detail("Unsupported scenario, " + unsupported_scenario);
+    });
+
+    RunPerformanceMode<scenario>(sut, qsl, base_settings, sequence_gen);
+    return;
+  }
+
+  LogDetail([base_field =
+                 find_peak_performance::ToStringPerformanceField<scenario>(
+                     base_settings)](AsyncDetail& detail) {
+    detail("FindPeakPerformance: Check validity of the base settings field: " +
+           base_field);
+  });
+
+  std::vector<loadgen::LoadableSampleSet> base_loadable_sets(
+      loadgen::GenerateLoadableSets(qsl, base_settings));
+  const LoadableSampleSet& base_performance_set = base_loadable_sets.front();
+  LoadSamplesToRam(qsl, base_performance_set.set);
+
+  PerformanceResult base_pr(IssueQueries<scenario, TestMode::PerformanceOnly>(
+      sut, base_settings, base_performance_set, sequence_gen));
+  PerformanceSummary base_perf_summary{sut->Name(), base_settings,
+                                       std::move(base_pr)};
+
+  // Check whether base settings satisfy the performance constraints or
+  // not.
+  std::string msg;
+  if (!base_perf_summary.PerfConstraintsMet(&msg)) {
+    LogDetail([msg](AsyncDetail& detail) {
+      detail(
+          "FindPeakPerformance: Initial lower bound does not satisfy "
+          "performance constraints, msg: " +
+          msg);
+    });
+
+    sut->ReportLatencyResults(base_perf_summary.pr.latencies);
+
+    LogSummary(
+        [perf_summary = PerformanceSummary{sut->Name(), base_settings,
+                                           std::move(base_perf_summary.pr)}](
+            AsyncSummary& summary) mutable { perf_summary.Log(summary); });
+
+    qsl->UnloadSamplesFromRam(base_performance_set.set);
+
+    return;
```

Comment:
Can you do a search on the left boundary, like you do for the right boundary?

It might help if there was an additional TestSetting:
double server_find_peak_qps_boundary_step_size = 1;

For the right boundary, you'd grow it by:
server_target_qps = server_target_qps *  (1 + server_find_peak_qps_boundary_step_size);

Then, for the left boundary, you'd grow it by:
server_target_qps = server_target_qps /  (1 + server_find_peak_qps_boundary_step_size));

Reply:
For this patch, let's not worry about expanding the left boundary. However, can you add a comment to one of the TestSettings fields about the initial server_target_qps being used as the lower bound?